### PR TITLE
Mobilecoind db encryption: Take 2

### PR DESCRIPTION
**Pull request by eranrund**
Note: This is still missing tests and self-review, but I wanted to put it out there in case anyone wants to start looking at it.

### Motivation

A new feature request was introduced after the original PR (#577) was created: We need to be able to change the password, which means re-encrypting the data (or transitioning from no password to having a password, or the other way around).

### In this PR
Re-write the mobilecoind encryption solution. The major changes are:
1) We now store a test key/value pair inside the LMDB database. This allows the crypto code to check if encryption is enabled or not: If the key is present, then encryption is enabled and the password can be verified by attempting to decrypt the value of this test key and seeing if it matches the hardcoded data we expect it to have. When no encryption is enabled, this key is not set. This allows this change to be backwards-compatible, and removes the need to explicitly specify `--encrypted-db` via command line.
2) The plug-able encryption trait thing was removed, since now you can switch between encrypted and unencrypted modes during runtime. The database starts empty and un-encrypted. Setting a password would create the test key mentioned above, as well as encrypt all the MonitorStore data with the new password. If the password is set to empty (an empty vector), no encryption actually takes place - this allows the work here to be backwards compatible.
3) Support was added for re-encrypting previously encrypted data, which allows changing the password after it has been set.
4) API call changes: We now have two API calls for password management:
    - SetDbPassword: This sets a new password. It will only succeed if either no password has previously been set, or if the database has been unlocked by UnlockDb
    - UnlockDb: Provides the password for decrypting a previously-encrypted database. This can only be called once during the lifetime of a mobilecoind invocation (since it makes no sense to do it multiple times).

### CODE REVIEW

File: mobilecoind/src/db_crypto.rs [hunk 1]
```diff
@@ -11,33 +11,44 @@ use aes_gcm::{
...
 /// This is set to 32 bytes as the intended purpose is for the user to pass a hash of a
 /// password and not the actual password the user typed.
-pub const AES_PASSWORD_LEN: usize = 32;
+pub const PASSWORD_LEN: usize = 32;
+
+/// LMDB database name for storing metadata.
+const CRYPTO_DB_NAME: &str = "db_crypto";
+
+/// Key/value used for testing we have the correct encryption key.
+const TEST_KEY: &str = "encryption test key";
```
Comment (sugargoat) [inline mobilecoind/src/db_crypto.rs:32]: Naming nit - I think this is a little confusing as named. Maybe would suggest 

```rust
const ENCRYPTION_STATE_KEY: &str = "db_encrypted";
const ENCRYPTION_STATE_VAL: &str = "true";
```

File: mobilecoind/src/db_crypto.rs [hunk 2]
```diff
@@ -46,139 +57,241 @@ impl From<AeadError> for DbCryptoError {
...
+impl DbCryptoProvider {
+    pub fn new(env: Arc<Environment>) -> Result<Self, DbCryptoError> {
+        let database = env.create_db(Some(CRYPTO_DB_NAME), DatabaseFlags::empty())?;
+
+        // Check if the database is currently encrypted.
+        let is_db_encrypted = {
+            let db_txn = env.begin_ro_txn()?;
+            match db_txn.get(database, &TEST_KEY.as_bytes()) {
+                Ok(_test_val) => {
+                    // The test key is present in the database, this means encryption is enabled.
```
Comment (sugargoat) [inline mobilecoind/src/db_crypto.rs:99]:
```suggestion
                    // The encryption indicator key is present in the database, this means encryption is enabled.
```

File: mobilecoind/src/db_crypto.rs [hunk 3]
```diff
@@ -46,139 +57,241 @@ impl From<AeadError> for DbCryptoError {
...
+                    true
+                }
+                Err(LmdbError::NotFound) => {
+                    // The test key is not in the database, this means encryption is not enabled.
```
Comment (sugargoat) [inline mobilecoind/src/db_crypto.rs:103]:
```suggestion
                    // The encryption indicator key is not in the database, this means encryption is not enabled.
```

File: mobilecoind/src/db_crypto.rs [hunk 4]
```diff
@@ -46,139 +57,244 @@ impl From<AeadError> for DbCryptoError {
...
-    }
-}
+        // There are two scenarios in which a password won't be set:
+        // 1) When the db is not being encrypted, in which case we just return the
+        //    ciphertext back to the caller.
+        // 2) When the db is encrypted but has not yet been unlocked by calling
+        //    check_and_store_password. In order to provide a better user experience, we
+        //    test if that is the case before assuming no password is required.
+        //    This allows callers to get a meaningful error (PasswordNeeded) instead of
+        //    prost decode errors.
```
Comment (sugargoat) [inline mobilecoind/src/db_crypto.rs:266]: Nice

File: mobilecoind/src/db_crypto.rs [hunk 5]
```diff
@@ -46,139 +57,244 @@ impl From<AeadError> for DbCryptoError {
...
+        match (state.is_db_encrypted, state.encryption_key.is_empty()) {
+            // Db is not encrypted and password is empty
+            (false, true) => Ok(ciphertext.to_vec()),
+
+            // Db is not encrypted and password is not empty (should never happen)
+            (false, false) => panic!("invalid state"),
```
Comment (sugargoat) [inline mobilecoind/src/db_crypto.rs:272]:
```suggestion
            (false, false) => panic!("invalid db encryption state"),
```

File: mobilecoind/src/service.rs [hunk 6]
```diff
@@ -1570,20 +1559,51 @@ impl<
...
-            .set_password(request.get_password())
-            .map_err(|err| rpc_internal_error("set_db_password", err, &self.logger))?;
-
-        // Get the monitors list - this will attempt to decode any potential monitors, and will
-        // fail if the password is incorrect. If there are no monitors then we have no encrypted
-        let _ = self.mobilecoind_db.get_monitor_map().map_err(|err| {
-            let _ = self.crypto_provider.clear_password();
-            rpc_internal_error("set_db_password.validate", err, &self.logger)
-        })?;
+        // Check if the database is unlocked and allowing this operations.
```
Comment (sugargoat) [inline mobilecoind/src/service.rs:1562]:
```suggestion
        // Check if the database is unlocked and allowing this operation.
```

File: mobilecoind/src/service.rs [hunk 7]
```diff
@@ -1570,20 +1559,51 @@ impl<
...
+        }
+
+        self.mobilecoind_db
+            .check_and_store_password(&request.get_password())
+            .map_err(|err| {
+                rpc_internal_error("mobilecoind_db.check_and_store_password", err, &self.logger)
+            })?;
+
+        log::info!(self.logger, "Successfully unlocked, starting sync thread.");
+        (self.start_sync_thread)();
```
Comment (sugargoat) [inline mobilecoind/src/service.rs:1606]: 👍 I like the way the syncing is gated on unlocking.

File: start-testnet-client.sh [hunk 8]
```diff
@@ -16,19 +16,11 @@ pushd "$(dirname "$0")"
 
 echo "Pulling down TestNet consensus validator signature material"
 
-SIGSTRUCT_URI=$(curl -s https://enclave-distribution.test.mobilecoin.com/production.json | grep sigstruct | awk '{print $2}' | tr -d \")
-curl -O https://enclave-distribution.test.mobilecoin.com/${SIGSTRUCT_URI}
-
-TARGETDIR=./target/release
+TARGETDIR=./target/debug
```
Comment (sugargoat) [inline start-testnet-client.sh:19]: Were these changes intentional?

**Review by sugargoat — COMMENTED** [review]
Looking great

File: start-testnet-client.sh (hunk 8 above)
Comment (eranrund) [reply start-testnet-client.sh:19]: Definitely not

**Review by sugargoat — APPROVED** [review]
LGTM

**Review by tsegaran — APPROVED** [review]
This looks good, feel free to merge.
